internal/registry/registry.go

```diff
@@ -245,3 +259,124 @@ func (r *registry) getImageDigestFromMultiImage(manifestListStream []byte) (stri
 	}
 	return "", fmt.Errorf("Failed to find manifest for architecture %s", arch)
 }
+
+func (r *registry) GetImageByName(imageName string, auth authn.Authenticator) (v1.Image, error) {
+
+	ref, err := r.ParseReference(imageName)
+	if err != nil {
+		return nil, err
+	}
+
+	descriptor, err := remote.Get(ref, remote.WithAuth(auth))
+	if err != nil {
+		return nil, fmt.Errorf("could not get image: %w", err)
+	}
+
+	img, err := descriptor.Image()
+	if err != nil {
+		return nil, fmt.Errorf("could not call image: %w", err)
+	}
+	return img, nil
+}
+
+func (r *registry) ParseReference(imageName string) (name.Reference, error) {
+	opts := make([]name.Option, 0)
+	ref, err := name.ParseReference(imageName, opts...)
+	if err != nil {
+		return nil, fmt.Errorf("could not parse the container image %s: %w", imageName, err)
+	}
+
+	return ref, nil
+}
+
+func (r *registry) WriteImageByName(imageName string, image v1.Image, auth authn.Authenticator) error {
+
+	ref, err := r.ParseReference(imageName)
+	if err != nil {
+		return err
+	}
+
+	err = remote.Write(ref, image, remote.WithAuth(auth))
+	if err != nil {
+		return fmt.Errorf("failed to push signed image: %w", err)
+	}
+	return nil
+}
+
+func (r *registry) AddLayerToImage(tarfile string, image v1.Image) (v1.Image, error) {
+
+	//turn our tar archive into a layer
+	mt, err := r.GetLayerMediaType(image)
+	if err != nil {
+		return nil, err
+	}
+
+	signedLayer, err := tarball.LayerFromFile(tarfile, tarball.WithMediaType(mt))
+	if err != nil {
+		return nil, fmt.Errorf("failed to generate layer from tar: %w", err)
+	}
+
+	// add the layer to the unsigned image
+	newImage, err := mutate.AppendLayers(image, signedLayer)
+	if err != nil {
+		return nil, fmt.Errorf("failed to append layer: %w", err)
+	}
+
+	// this is needde because mutate.AppendLayers loses the image mediatype
+	// without it the controller refuses to run the resulting images
+	imageMediaType, _ := image.MediaType()
+	newImageWithMT := mutate.MediaType(newImage, imageMediaType)
+	if err != nil {
+		return nil, fmt.Errorf("failed to change medaitype of image: %w", err)
+	}
+	return newImageWithMT, nil
+}
+
+func (r *registry) GetLayerMediaType(image v1.Image) (types.MediaType, error) {
+	layers, err := image.Layers()
+	if err != nil {
+		return types.OCILayer, fmt.Errorf("could not get the layers from image: %w", err)
+	}
+	return layers[len(layers)-1].MediaType()
+}
+
+/*
+** generic function to loop through all the files in an image and run
+** a function on them, based loosly on ftw() or filepath.Walk().
+** image  - the image to examine
+** fn	  - the function to call on each file
+** data[] - an array of strings to pass through to the helper function
+**        to provide any additional data needed.
+ */
+func (r *registry) WalkFilesInImage(image v1.Image, fn func(filename string, header *tar.Header, tarreader io.Reader, data []string) error, data []string) error {
```

Comment:
This function walk on all the files in all the layers. We are only interested in the file under specific directory. So maybe we should add an option to this function or to processing function. WDYT?

Reply:
KMMO is only interested in the files under specific directory, but I am trying to keep this image as generic as possible so it is still usable outside KMMO. That means we can't assume that all the kmods are in a single directory, so it would have to be an array (or similar) of directories and that adds a lot more complication.  Plus then future users may have different requirements.

The signimage job currently runs through in a few tens of seconds, so I don't really think there's much need to optomise this function for performance at the moment.

So, maybe in the future, but I dont think we need to do that in this PR.